File: 2021/12/solve.py
#!/usr/bin/python
import os
import sys

path = os.path.join(os.path.dirname(__file__), sys.argv[1] if len(sys.argv) > 1 else 'input')

adjacent = {}
for line in open(path, 'r'):
    cave1, cave2 = line.strip().split('-')

    if cave1 != 'end' and cave2 != 'start':
        if cave1 in adjacent:
            adjacent[cave1].add(cave2)
        else:
            adjacent[cave1] = {cave2}

    if cave2 != 'end' and cave1 != 'start':
        if cave2 in adjacent:
            adjacent[cave2].add(cave1)
        else:
            adjacent[cave2] = {cave1}

answer1 = 0
answer2 = 0

stack = [('start', set(), True)]
while stack:
    cave, visited, allow_second_visit = stack.pop()

    if cave.islower():
        visited = {*visited, cave}

    for adjacent_cave in adjacent[cave]:
        if adjacent_cave == 'end':
            if allow_second_visit:
                answer1 += 1
            answer2 += 1
        elif adjacent_cave not in visited:
            stack.append((adjacent_cave, visited, allow_second_visit))
        elif allow_second_visit:
            stack.append((adjacent_cave, visited, False))

print('--- Day 12: Passage Pathing ---')
print(' Answer 1: ' + str(answer1))
print(' Answer 2: ' + str(answer2))

File: 2022/12/solve.r
#!/usr/bin/env Rscript

all_args <- commandArgs(trailingOnly=FALSE)
args <- commandArgs(trailingOnly=TRUE)
if (all_args[1] == 'RStudio') {
	library('rstudioapi')
	path <- rstudioapi::getSourceEditorContext()$path
} else {
	path <- tail(c('.', gsub('^(?:|--file=)(.*)$', '\\1', all_args[c(which(all_args == '-f') + 1, which(grepl('^--file=', all_args)))])), 1)
}
path <- file.path(normalizePath(dirname(path)), ifelse(length(args)>0, args[1], 'input'))

map <- NULL
S1 <- list()
S2 <- list()
E <- NULL

R <- 0L
for (line in strsplit(readLines(path), '')) {
	R <- R + 1L
	if (is.null(map)) {
		map <- matrix(nrow=0, ncol=length(line))
	}
	row = integer()

	C <- 0L
	for (char in line) {
		C <- C + 1L
		if (char == 'S') {
			row <- c(row, 1L)
			S1 <- c(S1, list(c(R, C)))
			S2 <- c(S2, list(c(R, C)))
		} else if (char == 'a') {
			row <- c(row, 1L)
			S2 <- c(S2, list(c(R, C)))
		} else if (char == 'E') {
			row <- c(row, 26L)
			E <- c(R, C)
		} else {
			row <- c(row, which(char == letters))
		}
	}
	map <- rbind(map, c(row))
}

calculate <- function(start) {
	distance <- 0L

	queue <- list()
	for (coordinates in start) {
		queue <- c(queue, list(c(coordinates, 0L))) # (row, col, distance)
	}
	rows <- nrow(map)
	cols <- ncol(map)
	visited <- character()

	while (length(queue) > 0) {
		item <- queue[[1]]
		queue <- tail(queue, -1)

		if (!(paste(item[1:2], collapse=',') %in% visited)) {
			visited <- c(visited, paste(item[1:2], collapse=','))
			if (item[1] == E[1] & item[2] == E[2]) {
				distance <- item[3]
				break
			} else {
				max_height <- map[item[1], item[2]] + 1L
				for (R in c(item[1]-1L, item[1]+1L)) {
					if (R > 0 & R <= rows) {
						if (map[R, item[2]] <= max_height) {
							queue <- c(queue, list(c(R, item[2], item[3]+1L)))
						}
					}
				}
				for (C in c(item[2]-1L, item[2]+1L)) {
					if (C > 0 & C <= cols) {
						if (map[item[1], C] <= max_height) {
							queue <- c(queue, list(c(item[1], C, item[3]+1L)))
						}
					}
				}
			}
		}
	}

	distance
}

answer1 <- calculate(S1)
answer2 <- calculate(S2)

cat('--- Day 12: Hill Climbing Algorithm ---\n')
cat(' Answer 1:', answer1, '\n')
cat(' Answer 2:', answer2, '\n')
